locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/softlayer/SoftLayerSameVlanLocationCustomizer.java

```diff
@@ -0,0 +1,398 @@
+
+    /**
+     * Update the {@link org.jclouds.compute.options.TemplateOptions} that will
+     * be used by {@link JcloudsLocation} to obtain machines. Uses the VLAN
+     * numbers configured on an existing machine, as saved in the configuration
+     * maps for {@link #PUBLIC_VLAN_ID_MAP public} and {@link #PRIVATE_VLAN_ID_MAP private}
+     * VLAN numbers.
+     *
+     * @see {@link JcloudsLocationCustomizer#customize(JcloudsLocation, ComputeService, TemplateOptions)}
+     */
+    @Override
+    public void customize(JcloudsLocation location, ComputeService computeService, TemplateOptions templateOptions) {
+        // Check we are customising a SoftLayer location
+        String provider = location.getProvider();
+        if (!(provider.equals(SoftLayerConstants.SOFTLAYER_PROVIDER_NAME) &&
+                templateOptions instanceof SoftLayerTemplateOptions)) {
+            String message = String.format("Invalid location provider or template options: %s/%s",
+                    provider, templateOptions.getClass().getSimpleName());
+            LOG.warn(message);
+            throw new IllegalArgumentException(message);
+        }
+
+        // Check template options for VLAN configuration and save them if set
+        String scopeUid = getScopeUid(location);
+        SoftLayerTemplateOptions softLayerOptions = (SoftLayerTemplateOptions) templateOptions;
+        Integer publicVlanId = softLayerOptions.getPrimaryNetworkComponentNetworkVlanId();
+        Integer privateVlanId = softLayerOptions.getPrimaryBackendNetworkComponentNetworkVlanId();
+        if (publicVlanId != null || privateVlanId != null) {
+            LOG.debug("SoftLayer VLANs private {} and public {} already configured in template options for scope: {}",
+                    new Object[] { privateVlanId, publicVlanId, scopeUid });
+            savePublicVlanId(location, scopeUid, publicVlanId);
+            savePrivateVlanId(location, scopeUid, privateVlanId);
+            return;
+        }
+
+        // Create a new latch if we are first
+        CountDownLatch latch = null;
+        synchronized (lock) {
+            latch = lookupCountDownLatch(location, scopeUid);
+            if (latch == null) {
+                // We are first to try obtaining a VM; create a latch and return
+                LOG.debug("Creating new latch for scope: {}", scopeUid);
+                latch = createCountDownLatch(location, scopeUid);
+                return;
+            }
+        }
+
+        // Try and acquire the latch, or time out
+        Duration timeout = getTimeout(location);
+        Tasks.setBlockingDetails("Waiting for VLAN details");
+        try {
+            LOG.debug("Waiting for VLAN details for scope: {}", scopeUid);
+            if (!Uninterruptibles.awaitUninterruptibly(latch, timeout.toMilliseconds(), TimeUnit.MILLISECONDS)) {
+                // Release the latch and throw an exception to prevent others blocking forever
+                latch.countDown();
```

Comment:
My read is still that if A takes too long an B detects the timeout:

* B fails -- `throw` on line 246 -- causing the entity start to fail
* C and D return without setting any vlan info -- line 258 -- and then proceed to provision, and then the first of them will proceed to set these as the official info (line 184), whereas the other will quietly exit without even noticing it's possibly on a different subnet (line 163)

So A is still off and running, B has failed, and C and D may be on separate subnets.  However subsequent entities will correctly pick up the vlan's assigned to C.

Would a simple fix be to throw on line 258 (so C and D fail) but also for B to remove the latch in the map (so that a subsequent run call it E will generate a new latch and repeat the process; if in the meantime A succeeds then we'll nicely get its VLAN info; we live with the mess if A succeeds while E is provisioning, as long as we have a warning re E on line 161).

HTH